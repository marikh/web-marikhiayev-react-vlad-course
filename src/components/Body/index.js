import React from 'react';
import './body.css';

export default ({ children }) => (
    <section className="App-body">
        { children }
    </section>
)