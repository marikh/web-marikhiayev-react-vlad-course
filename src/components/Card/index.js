import React from 'react';
import './card.css';
import classNames from 'classnames';

export default ({ children, className: parentClassName, ...props }) => (
    <div className={classNames(['raised-2','card', parentClassName])} {...props}>
      { children }
    </div>
)