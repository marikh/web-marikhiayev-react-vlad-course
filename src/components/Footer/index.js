import React from 'react';
import './footer.css';

export default ({ children }) => (
    <footer className="App-footer">
        { children }
    </footer>
)

