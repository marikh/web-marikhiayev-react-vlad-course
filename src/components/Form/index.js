import React,{Component} from 'react';
import classNames from 'classnames';
import './form.css';
import { 
    Layout,
    Section,
    Heading,
    InputField
} from '../../components/';
import { Prompt } from 'react-router-dom';

class Form extends Component{

    constructor(props){
        super(props)

        this.updateForm = this.updateForm.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.state = { 
            formFields : this.props.formFields,
            submitted:false
        };
    }

    updateForm(name, value){

        const changedField = this.state.formFields.find(field => field.name === name);
        changedField.value = value;

        const indexOfItemToUpdate =  this.state.formFields.indexOf(changedField);
        const updatedFormFieldsArray = [...this.state.formFields.slice(0, indexOfItemToUpdate), changedField, ...this.state.formFields.slice(indexOfItemToUpdate + 1)];

        this.setState({ formFields : updatedFormFieldsArray , submitted: false});

    }

    
    onSubmit(e){
        e.preventDefault();

        this.resetFields();
        this.setState({ formFields : [...this.state.formFields] , submitted: true });        
    }

    resetFields(){
        for(let i=0; i<this.state.formFields.length; i++){
            this.state.formFields[i].value = '';
        }
    }

render(){
    return (
        <div className="form-div">
                <form onSubmit={this.onSubmit}>
                    <div className="form-group" style={{ padding: '2rem', display: 'flex', justifyContent: 'center', flexDirection: 'column' }}>
                    {  this.state.formFields.map(( {name, label, type, value, required} ) => (
                        <InputField key={`input-field-${name}`} name={name} 
                                    type={type} value={value} label={label} 
                                    required={required}
                                    updateForm={this.updateForm} />
                        ))
                    }
                    
                    <button type="submit">Submit</button>
                    </div>
                </form>
                <div className={classNames([this.state.submitted && 'submitted-notification'])}/>
                <Prompt
                    when={this.state.formFields.some(field => field.value !== '')}
                    message={location => (
                        `Are you sure you want to go to 
                            ${location.pathname} without submitting the form?`
                    )}
                />
        </div>
    )
    }
}

export default Form;



