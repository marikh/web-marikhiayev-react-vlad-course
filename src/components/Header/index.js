import React from 'react';
import Menu from '../Menu';

export default ({ children, menuConfig }) => (
    <header className="App-header">
        <Menu menuConfig={menuConfig} />
        <section className="App-subheader">{ children }</section>
    </header>
)