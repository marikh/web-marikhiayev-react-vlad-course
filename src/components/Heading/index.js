import React from 'react';
import classNames from 'classnames';
import './headings.css';

export default ({ size, children, className, ...props }) => 
    React.createElement(
        `h${size}`,{ 
            className: classNames([
                className, `heading heading-${size}`
            ]), ...props 
        } , children
    )