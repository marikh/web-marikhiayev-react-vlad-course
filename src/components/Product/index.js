import React from 'react';
import './product.css';
import classNames from 'classnames';
import T from 'i18n-react';
import {Card} from '../../components'
import {
  Link
} from 'react-router-dom'

const Product = ({ id, name, imageUrl, shortDesc, price, cartItemId, deleteProductFromCart }) => (
    <Card className="product-card" style={{ backgroundImage: `url(${imageUrl})`}}>
        <Link to={`/products/${id}`} className="product-wrapping-link">
            <div className="product-top-title">
                <p className="product-title">{ name }</p>
                <p className="product-price">{ price }</p>
            </div>
            <p className="product-bottom-desc">{ shortDesc }</p>
        </Link>
        { deleteProductFromCart && cartItemId && <button className="cart-product-delete-button" id="remove-button" onClick={(e) => deleteProductFromCart(cartItemId)}>{T.translate("Delete")}</button>}
    </Card>
)

export default Product;