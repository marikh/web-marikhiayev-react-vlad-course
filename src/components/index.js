import Layout from './Layout';
import Header from './Header/';
import Footer from './Footer/';
import Body from './Body/';
import HeroArea from './HeroArea/';
import SocialLinks from './SocialLinks/';
import Heading from './Heading/';
import Section from './Section/';
import Card from './Card/';
import Menu from './Menu/';
import InputField from './InputField/';

export {
    Layout,
    Header,
    Body,
    Footer,
    SocialLinks,
    HeroArea,
    Heading,
    Section,
    Card,
    Menu,
    InputField
}