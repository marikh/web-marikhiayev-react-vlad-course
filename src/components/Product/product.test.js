import { mount, shallow } from "enzyme";
import React from 'react';
import { MemoryRouter, Route, Link } from 'react-router-dom'
import { Card, Product } from '../../components/';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux';
import {createStore} from 'redux';

const testNonCartProduct =
    {
        id: '123123-234-2341-123123-123123',
        name: 'Greek amphora 1',
        imageUrl: '//c1.staticflickr.com/5/4215/35504896635_ec1a78af43_b.jpg',
        shortDesc: 'Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.',
        longDesc: 'Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut. Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.',
        price: '101K USD'
    };
    
const cartProduct =
    {
        cartItemId: 2,
        id: '123123-12342-456456-123123-123123',
        name: 'Greek amphora 2',
        imageUrl: '//c1.staticflickr.com/5/4215/35504896635_ec1a78af43_b.jpg',
        shortDesc: 'Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.',
        longDesc: 'Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut. Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.',
        price: '102K USD'
    };

const mockStore = configureStore()

const initialState = { } ;
    let store, wrapper;  

describe('Product outer tests', () => {

     beforeEach(()=>{
        if(wrapper != null)
            wrapper.unmount();

        store = mockStore(initialState);
        wrapper = mount( <Provider store={store}>
                            <MemoryRouter>
                                <Product {...cartProduct} deleteProductFromCart={() => {}}/>
                            </MemoryRouter>
                        </Provider> );
    })

    it("contains Card", () => {
        expect(wrapper.find(Card).length).toBeGreaterThan(0)
    })
    
    it("has link to product page", () => {
        expect(wrapper.find('a[href="/products/123123-12342-456456-123123-123123"]').length).toBe(1)
    })

    it("has product title", () => {
        expect(wrapper.contains(<p className="product-title">{cartProduct.name}</p>)).toBe(true);
    });

    it("has product price", () => {
        expect(wrapper.contains(<p className="product-price">{cartProduct.price}</p>)).toBe(true);
    });

    it("has product bottom description", () => {
        expect(wrapper.contains(<p className="product-bottom-desc">{cartProduct.shortDesc}</p>)).toBe(true);
    });

    it("remove button appears (since the wrapper instance is a cart product)", () => {
        expect(wrapper.find('#remove-button').length).toBe(1);
    });


    it("no remove button should appears (since the wrapper instance is not a cart product - no cartItemId is provided in the props)", () => {
        
        if(wrapper != null)
            wrapper.unmount();

        store = mockStore(initialState);
        wrapper = mount( <Provider store={store}>
                            <MemoryRouter>
                                <Product {...testNonCartProduct} />
                            </MemoryRouter>
                        </Provider> );
        
        
        expect(wrapper.find('#remove-button').length).toBe(0);
    });
})

describe('Product Shallow Render',()=>{
    let wrapper;

    beforeEach(()=>{
        wrapper = shallow(<Product />)
    })

    it('render the DUMB component', () => {
       expect(wrapper.length).toEqual(1)
    });
})


describe('Product snapshot tests', () => {

    it('test weird snapshot', () => {

        const store = mockStore(initialState);

        const tree = renderer.create(
            <Provider store={store}>
                            <MemoryRouter>
                                <Product {...cartProduct} deleteProductFromCart={() => {}}/>
                            </MemoryRouter>
                        </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    })
 })

