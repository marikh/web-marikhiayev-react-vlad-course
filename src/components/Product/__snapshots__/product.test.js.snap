// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product snapshot tests test weird snapshot 1`] = `
<div
  className="raised-2 card product-card"
  style={
    Object {
      "backgroundImage": "url(//c1.staticflickr.com/5/4215/35504896635_ec1a78af43_b.jpg)",
    }
  }
>
  <a
    className="product-wrapping-link"
    href="/products/123123-12342-456456-123123-123123"
    onClick={[Function]}
  >
    <div
      className="product-top-title"
    >
      <p
        className="product-title"
      >
        Greek amphora 2
      </p>
      <p
        className="product-price"
      >
        102K USD
      </p>
    </div>
    <p
      className="product-bottom-desc"
    >
      Cupidatat irure magna magna in Lorem Lorem est tempor tempor ut.
    </p>
  </a>
  <button
    className="cart-product-delete-button"
    id="remove-button"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
